tests/switchless_one_tcs/enc/enc.c

```diff
@@ -0,0 +1,30 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <openenclave/corelibc/string.h>
+#include <openenclave/enclave.h>
+#include <openenclave/internal/print.h>
+#include "switchless_one_tcs_t.h"
+
+void enc_empty_regular(void)
+{
+}
+
+void enc_empty_switchless(status_e* status)
+{
+    int num_waiting = 0;
+    *status = FUNC_WORKING;
+    while (*status != FUNC_EXIT)
```

Comment:
The compiler is again optimizing this loop to :
```c
while (1)
{
      num_waiting++;
}
```

There is nothing in the body that indicates that the value of `status` will change within the body. 
Here is a code snippet: https://godbolt.org/z/F5MPTq

Given 
```c
typedef int status_e;
static status_e switchless_func_status;

#define FUNC_WORKING 1
#define FUNC_EXIT 2

void enc_empty_switchless(status_e* status)
{
    int num_waiting = 0;
    *status = FUNC_WORKING;
    while (*status != FUNC_EXIT)
    {
        num_waiting++;
    }
}
```

the compiler optimizes this to:
```asm
enc_empty_switchless:                   # @enc_empty_switchless
        mov     dword ptr [rdi], 1
.LBB0_1:                                # =>This Inner Loop Header: Depth=1
        jmp     .LBB0_1
```

Reply:
The case ran timeout should be due to switchless enclave worker thread failed before calling the enc_empty_switchless function.

log detail:

> OneTCSTest::test Create enclave with Switchless Calls and one TCS only.
Call empty switchless ecall, it is expected to pass...
Call empty ecall, it is expected to fail since TCS is occupied by trusted worker thread...
2020-05-13T06:23:57.000000Z [(H)ERROR] tid(0x7f1769d88700) | :OE_OUT_OF_THREADS [/home/xy/openenclave/host/sgx/calls.c:oe_ecall:616]
2020-05-13T06:23:57.000000Z [(H)ERROR] tid(0x7f1769d88700) | :OE_OUT_OF_THREADS [/home/xy/openenclave/host/calls.c:oe_call_enclave_function_by_table_id:90]
2020-05-13T06:23:57.000000Z [(H)ERROR] tid(0x7f1769d88700) | Switchless enclave worker thread failed


Backtrace ：
> 
(gdb) bt
#0  syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
#1  0x00000000004258d0 in _worker_wait (event=0x6c88b4) at /home/xy/openenclave/host/sgx/linux/switchless.c:31
#2  0x0000000000425712 in oe_host_worker_wait (context=0x6c88a0)
    at /home/xy/openenclave/host/sgx/linux/switchless.c:55
#3  0x0000000000423260 in _switchless_ocall_worker (arg=0x6c88a0) at /home/xy/openenclave/host/sgx/switchless.c:63
#4  0x00007ffff74ee6db in start_thread (arg=0x7ffff5589700) at pthread_create.c:463
#5  0x00007ffff721788f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

Reply:
Thanks @yanxue-intel . Let me investigate and update this thread.

Reply:
@yanxue-intel Thanks for your help. The reason the latest version was sometimes failing was due to an order of initialization. Somethings before an ecall worker thread could enter an enclave, a regular ecall could already have taken up the sole tcs. Thus the ecall worker thread may fail to initialize. I made a code fix to carefully order the ecalls during initialization so that all the normal ecalls first complete, then we launch the ecall worker threads and then wait to make sure that the threads have entered the enclave and taken up the tcs.
This leads to predictable behavior for enclave initialization and switchless worker threads.

Qualifying it in: #3065